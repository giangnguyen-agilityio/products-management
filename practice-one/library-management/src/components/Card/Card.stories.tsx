import React from 'react'
import { type StoryFn, type Meta } from '@storybook/react'
import Card from './index'

export default {
  title: 'Example/Card',
  component: Card
} as Meta<typeof Card>

const Template: StoryFn<typeof Card> = (args) => <Card {...args} />

export const Default = Template.bind({})
Default.args = {
  book: {
    bookId: 'B1',
    image: 'src/assets/images/book-1.jpg',
    alt: 'Harry Potter and the Prisoner of Azkaban book cover',
    title: 'Harry Potter and the Prisoner of Azkaban',
    author: 'J.K.Rowling',
    description:
      'The story follows Harry Potter during his third year at Hogwarts School of Witchcraft and Wizardry, where he learns that Sirius Black, an infamous wizard and convicted murderer, has escaped from the wizarding prison of Azkaban. As Harry delves into the mystery, aided by his friends Ron and Hermione, they encounter time travel, terrifying creatures called Dementors, and the truth about Harry\'s own past. Filled with adventure, friendship, and the complexities of the wizarding world, this captivating script takes readers on a thrilling magical journey.',
    price: 10.99,
    availableQuantity: 20,
    totalQuantity: 50
  },
  onRent: () => {
    console.log('Rent clicked')
  },
  onEdit: () => {
    console.log('Edit clicked')
  },
  onDelete: () => {
    console.log('Delete clicked')
  }
}
