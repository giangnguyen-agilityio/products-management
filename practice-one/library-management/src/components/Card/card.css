.card-wrapper {
  position: relative;
  cursor: pointer;
  padding-top: 320px;
  border: 2px solid transparent;
  transition: all 0.8s ease-in-out;
  list-style: none;
}

.card-image-wrapper {
  position: absolute;
  left: 20px;
  top: 0;
  width: 100%;
}

.card-image {
  width: inherit;
  height: 550px;
  border-radius: 10px;
}

.card-info {
  width: 100%;
  height: 100%;
  border: 2px solid var(--bg-btn-color);
  padding: 250px 30px 30px;
  border-radius: 10px;
}

.card-title {
  height: 50px;
  font-size: 26px;
  font-family: var(--secondary-font);
  margin: 0 0 10px;
}

.card-detail {
  display: flex;
  justify-content: space-between;
  align-items: center;
  font-family: var(--secondary-font);
  height: 45px;
}

.card-author,
.card-price {
  font-size: 24px;
}

.card-title,
.card-author,
.card-price {
  overflow: hidden;
  text-overflow: ellipsis;
}

.card-total-quantity,
.card-available-quantity {
  font-size: 24px;
  color: var(--primary-text-color);
}

.card-available-quantity {
  color: var(--secondary-color);
}

.card-author {
  width: 70%;
}

.card-description {
  height: 580px;
  margin: 20px 0 10px;
  font-size: 20px;
  line-height: 23px;
  color: var(--secondary-color);
}

.card-controls {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.card-button {
  display: flex;
  gap: 10px;
}

.card-rent-btn,
.edit-btn,
.delete-btn {
  padding: 15px;
  font-size: 20px;
  text-transform: uppercase;
  border-radius: 50%;
  border: 1px solid var(--bg-btn-color);
  font-family: var(--secondary-font);
  background: unset;
  transition: all 0.3s ease-in-out;
  cursor: pointer;
}

.card-rent-btn {
  border-radius: 15px;
  color: var(--primary-color);
}

.card-rent-btn:hover {
  border-color: var(--primary-color);
}

.edit-btn {
  color: var(--tertiary-color);
}

.edit-btn:hover {
  border-color: var(--tertiary-color);
}

.delete-btn {
  color: var(--danger-color);
}

.delete-btn:hover {
  border-color: var(--danger-color);
}

.card-wrapper:hover .card-info {
  border: 2px solid var(--primary-color);
  border-radius: 10px;
}

@media only screen and (max-width: 1440px) {
  .card-wrapper {
    margin: 0 auto;
  }

  .card-image-wrapper {
    position: absolute;
    left: 20px;
    top: 0px;
  }

  .card-image {
    height: 500px;
  }

  .card-info {
    padding: 200px 20px 30px;
  }

  .card-description {
    height: 580px;
    line-height: 26px;
  }
}

/* Media Query for Laptop Devices */
@media only screen and (max-width: 1024px) {
  .card-image-wrapper {
    position: absolute;
    left: 25px;
    top: 0px;
  }

  .card-image {
    height: 450px;
  }

  .card-info {
    padding: 150px 20px 30px;
  }

  .card-author,
  .card-price,
  .card-total-quantity,
  .card-available-quantity {
    font-size: 22px;
  }

  .card-description {
    height: 665px;
    line-height: 24px;
    margin: 20px 0;
    color: var(--secondary-color);
  }
}

/* Media Query for Tablets and Smaller Screens */
@media only screen and (max-width: 768px) {
  .card-image-wrapper {
    left: 0;
    width: 100%;
    padding: 0 5%;
  }

  .card-image {
    height: 450px;
  }

  .card-info {
    padding: 140px 30px 30px;
  }

  .card-description {
    height: 100%;
    font-size: 18px;
  }

  .card-controls {
    justify-content: center;
  }
}

/* Media Query for Mobile Devices */
@media only screen and (max-width: 480px) {
  .card-wrapper {
    padding: 0;
    margin: 0 auto;
  }

  .edit-btn,
  .delete-btn {
    padding: 10px;
  }

  .card-image-wrapper {
    position: unset;
  }

  .card-image {
    height: 350px;
  }

  .card-info {
    padding: 25px 10px 10px;
  }

  .card-title {
    height: unset;
    font-size: 20px;
  }

  .card-detail {
    height: unset;
    margin: 10px 0 0;
  }

  .card-author,
  .card-price,
  .card-total-quantity,
  .card-available-quantity {
    font-size: 17px;
  }

  .card-description {
    font-size: 15px;
    margin: 10px 0;
  }
}

/* Media Query for Old Mobile Devices */
@media only screen and (max-width: 320px) {
  .card-title {
    font-size: 20px;
  }

  .card-author,
  .card-price,
  .card-total-quantity,
  .card-available-quantity {
    font-size: 16px;
  }

  .card-info {
    padding: 10px;
  }

  .card-image {
    height: 250px;
  }

  .card-description {
    font-size: 15px;
    margin: 10px 0;
    line-height: 20px;
  }
}
