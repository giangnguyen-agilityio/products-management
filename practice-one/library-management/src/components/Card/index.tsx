import React, {memo} from 'react'
import Typography from '@components/commons/Typography'
import Button from '@components/commons/Button'
import {HiOutlinePencilSquare} from 'react-icons/hi2'
import {BsTrash3} from 'react-icons/bs'
import {IBook} from '@types'
import './card.css'

interface CardProps {
  book: IBook | null
  onRent: (id: string) => void
  onEdit: (id: string) => void
  onDelete: (id: string) => void
  isAdmin: boolean
}

const Card: React.FC<CardProps> = props => {
  // Destructure the props
  const {book, onRent, onEdit, onDelete, isAdmin} = props

  const handleRentBook = () => {
    onRent(book!.id)
  }

  const handleEditBook = () => {
    onEdit(book!.id)
  }

  const handleDeleteBook = () => {
    onDelete(book!.id)
  }

  return (
    <li className="card-wrapper">
      {/* Display the card image */}
      <div className="card-image-wrapper">
        {book?.image && (
          <img src={book?.image} alt={book?.title} className="card-image" />
        )}
      </div>

      {/* Display the card info */}
      <div className="card-info">
        {/* Book title */}
        <Typography variant="h3" className="card-title">
          {book?.title || 'This is the book title'}
        </Typography>

        {/* Book details (author and price) */}
        <div className="card-detail">
          <Typography variant="p" className="card-author">
            {book?.author || 'This is the book author'}
          </Typography>
          <Typography variant="p" className="card-price">
            {book?.price ? book.price.toFixed(2) : 0}
          </Typography>
        </div>

        {/* Book quantity */}
        <div className="card-detail">
          <Typography variant="p" className="card-total-quantity">
            Available quantity:
          </Typography>
          <Typography variant="p" className="card-available-quantity">
            {book?.availableQuantity ? book.availableQuantity : 0}/
            <span className="card-total-quantity">
              {book?.totalQuantity ? book.totalQuantity : 0}
            </span>
          </Typography>
        </div>

        {/* Book description */}
        <Typography variant="p" className="card-description">
          {book?.description || 'This is the book description'}
        </Typography>

        {/* Card controls */}
        <div className="card-controls">
          {/* Rent button */}
          {!isAdmin && (
            <Button
              className="card-rent-btn"
              ariaLabel="Card rent button"
              onClick={handleRentBook}
            >
              Rent
            </Button>
          )}

          {/* Edit and Delete buttons */}
          {isAdmin && (
            <div className="card-button">
              <Button
                className="edit-btn"
                ariaLabel="Edit button"
                onClick={handleEditBook}
              >
                <HiOutlinePencilSquare size={25} />
              </Button>
              <Button
                className="delete-btn"
                ariaLabel="Delete button"
                onClick={handleDeleteBook}
              >
                <BsTrash3 size={25} />
              </Button>
            </div>
          )}
        </div>
      </div>
    </li>
  )
}

export default memo(Card)
