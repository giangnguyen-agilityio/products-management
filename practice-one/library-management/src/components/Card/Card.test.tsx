import React from 'react'
import {render, screen, fireEvent} from '@testing-library/react'
import Card from '.'

const mockOnRent = jest.fn()
const mockOnEdit = jest.fn()
const mockOnDelete = jest.fn()

const mockBook = {
  id: '1',
  title: 'Mock Book Title',
  author: 'Mock Author',
  price: 10.99,
  description: 'Mock description',
  availableQuantity: 5,
  totalQuantity: 10,
  image: 'http://Book_image.example.com',
  alt: 'This is the book image',
}

describe('Card component', () => {
  it('should render the book details correctly', () => {
    const {container} = render(
      <Card
        book={mockBook}
        onRent={mockOnRent}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        isAdmin={false}
      />
    )

    expect(container).toMatchSnapshot()
  })

  it('should renders the default data when book data is not provided', () => {
    const {container} = render(
      <Card
        book={null}
        onRent={mockOnRent}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        isAdmin={true}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should render "Rent" button if not an admin', () => {
    const {container} = render(
      <Card
        book={mockBook}
        onRent={mockOnRent}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        isAdmin={false}
      />
    )
    const cardControls = container.querySelector('.card-controls')

    expect(screen.getByText(/Rent/i)).toBeInTheDocument()
    expect(cardControls).toMatchSnapshot()
  })

  it('should call onRent function when "Rent" button is clicked', () => {
    render(
      <Card
        book={mockBook}
        onRent={mockOnRent}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        isAdmin={false}
      />
    )

    fireEvent.click(screen.getByLabelText(/Card rent button/i))
    expect(mockOnRent).toHaveBeenCalledWith(mockBook.id)
  })

  it('should render "Edit" and "Delete" buttons if isAdmin is true', () => {
    const {container} = render(
      <Card
        book={mockBook}
        onRent={mockOnRent}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        isAdmin={true}
      />
    )
    const cardControls = container.querySelector('.card-controls')

    expect(screen.getByLabelText(/Edit button/i)).toBeInTheDocument()
    expect(screen.getByLabelText(/Delete button/i)).toBeInTheDocument()
    expect(cardControls).toMatchSnapshot()
  })

  it('should call onEdit function when "Edit" button is clicked', () => {
    render(
      <Card
        book={mockBook}
        onRent={mockOnRent}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        isAdmin={true}
      />
    )

    fireEvent.click(screen.getByLabelText(/Edit button/i))
    expect(mockOnEdit).toHaveBeenCalledWith(mockBook.id)
  })

  it('should call onDelete function when "Delete" button is clicked', () => {
    render(
      <Card
        book={mockBook}
        onRent={mockOnRent}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        isAdmin={true}
      />
    )

    fireEvent.click(screen.getByLabelText(/Delete button/i))
    expect(mockOnDelete).toHaveBeenCalledWith(mockBook.id)
  })

  it('should renders image alt attribute correctly with book title', () => {
    const {getByAltText} = render(
      <Card
        book={mockBook}
        onRent={mockOnRent}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
        isAdmin={true}
      />
    )
    const image = getByAltText(mockBook.title)
    expect(image).toBeInTheDocument()
  })
})
