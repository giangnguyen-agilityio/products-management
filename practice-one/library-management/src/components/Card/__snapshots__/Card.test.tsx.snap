// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should render "Edit" and "Delete" buttons if isAdmin is true 1`] = `
<div
  class="card-controls"
>
  <div
    class="card-button"
  >
    <button
      aria-label="Edit button"
      class="btn btn-primary btn-medium edit-btn"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="none"
        height="25"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="25"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16.862 4.487l1.687-1.688a1.875 1.875 0 112.652 2.652L10.582 16.07a4.5 4.5 0 01-1.897 1.13L6 18l.8-2.685a4.5 4.5 0 011.13-1.897l8.932-8.931zm0 0L19.5 7.125M18 14v4.75A2.25 2.25 0 0115.75 21H5.25A2.25 2.25 0 013 18.75V8.25A2.25 2.25 0 015.25 6H10"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <button
      aria-label="Delete button"
      class="btn btn-primary btn-medium delete-btn"
      type="button"
    >
      <svg
        fill="currentColor"
        height="25"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="25"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.5 1h3a.5.5 0 0 1 .5.5v1H6v-1a.5.5 0 0 1 .5-.5ZM11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3A1.5 1.5 0 0 0 5 1.5v1H2.506a.58.58 0 0 0-.01 0H1.5a.5.5 0 0 0 0 1h.538l.853 10.66A2 2 0 0 0 4.885 16h6.23a2 2 0 0 0 1.994-1.84l.853-10.66h.538a.5.5 0 0 0 0-1h-.995a.59.59 0 0 0-.01 0H11Zm1.958 1-.846 10.58a1 1 0 0 1-.997.92h-6.23a1 1 0 0 1-.997-.92L3.042 3.5h9.916Zm-7.487 1a.5.5 0 0 1 .528.47l.5 8.5a.5.5 0 0 1-.998.06L5 5.03a.5.5 0 0 1 .47-.53Zm5.058 0a.5.5 0 0 1 .47.53l-.5 8.5a.5.5 0 1 1-.998-.06l.5-8.5a.5.5 0 0 1 .528-.47ZM8 4.5a.5.5 0 0 1 .5.5v8.5a.5.5 0 0 1-1 0V5a.5.5 0 0 1 .5-.5Z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`Card component should render "Rent" button if not an admin 1`] = `
<div
  class="card-controls"
>
  <button
    aria-label="Card rent button"
    class="btn btn-primary btn-medium card-rent-btn"
    type="button"
  >
    Rent
  </button>
</div>
`;

exports[`Card component should render the book details correctly 1`] = `
<div>
  <li
    class="card-wrapper"
  >
    <div
      class="card-image-wrapper"
    >
      <img
        alt="Mock Book Title"
        class="card-image"
        src="http://Book_image.example.com"
      />
    </div>
    <div
      class="card-info"
    >
      <h3
        class="text-h3 text-default card-title"
      >
        Mock Book Title
      </h3>
      <div
        class="card-detail"
      >
        <p
          class="text-p text-default card-author"
        >
          Mock Author
        </p>
        <p
          class="text-p text-default card-price"
        >
          10.99
        </p>
      </div>
      <div
        class="card-detail"
      >
        <p
          class="text-p text-default card-total-quantity"
        >
          Available quantity:
        </p>
        <p
          class="text-p text-default card-available-quantity"
        >
          5
          /
          <span
            class="card-total-quantity"
          >
            10
          </span>
        </p>
      </div>
      <p
        class="text-p text-default card-description"
      >
        Mock description
      </p>
      <div
        class="card-controls"
      >
        <button
          aria-label="Card rent button"
          class="btn btn-primary btn-medium card-rent-btn"
          type="button"
        >
          Rent
        </button>
      </div>
    </div>
  </li>
</div>
`;
